File: Practica_1_2/Practica_1_2.m
%% Configuración de archivos y parámetros
textFiles = {'espanol.txt', 'ingles.txt', 'portugues.txt'};
languages = {'Español', 'Inglés', 'Portugués'};

% Ordenes de extensión a analizar (agrupación de n palabras)
orders = [1, 2, 3, 4];

% Matriz para almacenar las entropías: filas para cada idioma, columnas para cada orden
entropyMatrix = zeros(length(textFiles), length(orders));

%% Procesamiento para cada archivo
for fileIdx = 1:length(textFiles)
    filename = textFiles{fileIdx};
    lang = languages{fileIdx};
    
    % Leer el contenido del archivo
    textContent = fileread(filename);
    
    % Convertir a minúsculas (para evitar duplicados por mayúsculas)
    textContent = lower(textContent);
    
    % Separar en palabras (suponiendo que solo hay espacios como separador)
    words = strsplit(textContent);
    % Eliminar celdas vacías (por si hay espacios extra)
    words = words(~cellfun('isempty', words));
    
    fprintf('=========================================\n');
    fprintf('Archivo: %s (Idioma: %s)\n', filename, lang);
    
    %% Para cada extensión de orden n (agrupación de n palabras)
    for orderIdx = 1:length(orders)
        n = orders(orderIdx);
        
        % Crear los n-gramas (agrupaciones consecutivas)
        nGrams = cell(1, length(words)-n+1);
        for i = 1:(length(words)-n+1)
            nGrams{i} = strjoin(words(i:i+n-1), ' ');
        end
        
        % Contar ocurrencias únicas usando unique y accumarray
        [uniqueNGrams, ~, idx] = unique(nGrams);
        counts = accumarray(idx, 1);
        totalNGrams = sum(counts);
        
        % Calcular las probabilidades de cada n-grama
        probabilities = counts / totalNGrams;
        
        % Calcular la información teórica para cada n-grama
        information = -log2(probabilities);
        
        % Calcular la entropía de la fuente extendida
        entropyVal = sum(probabilities .* information);
        entropyMatrix(fileIdx, orderIdx) = entropyVal;
        
        % Ordenar los n-gramas según su probabilidad (de mayor a menor)
        [sortedProbs, sortIdx] = sort(probabilities, 'descend');
        top5NGrams = uniqueNGrams(sortIdx(1:min(5,end)));
        top5Probs = sortedProbs(1:min(5,end));
        top5Info = -log2(top5Probs);  % Equivalente a information para esos
        
        % Mostrar resultados en consola
        fprintf('\nOrden (agrupación de %d palabras):\n', n);
        fprintf('Entropía: %.4f bits por n-grama\n', entropyVal);
        fprintf('Top 5 n-gramas con mayor probabilidad:\n');
        for j = 1:length(top5NGrams)
            fprintf('%d. "%s" -> Probabilidad: %.8f, Información: %.4f bits\n', ...
                j, top5NGrams{j}, top5Probs(j), top5Info(j));
        end
        fprintf('-----------------------------------------\n');
    end
end

%% Comparación gráfica de entropías
% Cada fila corresponde a un idioma y cada columna a un orden (n)
figure;
bar(categorical(languages), entropyMatrix)
xlabel('Idioma')
ylabel('Entropía (bits/n-grama)')
title('Comparación de Entropía de fuentes extendidas')
legend("Orden 1", "Orden 2", "Orden 3", "Orden 4", 'Location', 'Best')
grid on;

File: Practica_1_2/Practica_1_2.py
import os
import numpy as np
import matplotlib.pyplot as plt
from collections import Counter
import math

# Configuración de archivos y parámetros
text_files = ['espanol.txt', 'ingles.txt', 'portugues.txt']
languages = ['Español', 'Inglés', 'Portugués']
orders = [1, 2, 3, 4]  # Ordenes: 1 = palabras individuales, 2 = bigramas, etc.

# Matriz para almacenar las entropías: filas para cada idioma, columnas para cada orden
entropy_matrix = np.zeros((len(text_files), len(orders)))

for file_idx, filename in enumerate(text_files):
    lang = languages[file_idx]
    
    # Leer el contenido del archivo (se asume codificación UTF-8)
    with open(filename, 'r', encoding='utf-8') as f:
        text_content = f.read()
    
    # Convertir a minúsculas
    text_content = text_content.lower()
    
    # Separar en palabras (asumiendo espacios como separador)
    words = text_content.split()
    # Eliminar posibles entradas vacías
    words = [w for w in words if w]
    
    print("=========================================")
    print(f"Archivo: {filename} (Idioma: {lang})")
    
    # Procesamiento para cada orden de extensión
    for order_idx, n in enumerate(orders):
        # Crear los n-gramas (agrupaciones consecutivas de n palabras)
        ngrams = [" ".join(words[i:i+n]) for i in range(len(words)-n+1)]
        
        # Contar ocurrencias únicas de cada n-grama
        counts = Counter(ngrams)
        total_ngrams = sum(counts.values())
        
        # Calcular la entropía y guardar la información de cada n-grama
        entropy_val = 0.0
        ngram_info = {}
        for ngram, count in counts.items():
            p = count / total_ngrams
            info = -math.log2(p)
            ngram_info[ngram] = (p, info)
            entropy_val += p * info
        entropy_matrix[file_idx, order_idx] = entropy_val
        
        # Obtener los 5 n-gramas con mayor probabilidad (ordenados de mayor a menor)
        sorted_ngrams = sorted(counts.items(), key=lambda x: x[1], reverse=True)
        top5 = sorted_ngrams[:5]
        
        print(f"\nOrden (agrupación de {n} palabra{'s' if n > 1 else ''}):")
        print(f"Entropía: {entropy_val:.4f} bits por n-grama")
        print("Top 5 n-gramas con mayor probabilidad:")
        for j, (ngram, count) in enumerate(top5, start=1):
            p = count / total_ngrams
            info = -math.log2(p)
            print(f"{j}. \"{ngram}\" -> Probabilidad: {p:.8f}, Información: {info:.4f} bits")
        print("-----------------------------------------")

# Comparación gráfica de entropías
x = np.arange(len(languages))
width = 0.2  # ancho de cada barra

fig, ax = plt.subplots()
for i, n in enumerate(orders):
    ax.bar(x + i * width, entropy_matrix[:, i], width, label=f"Orden {n}")

ax.set_xlabel("Idioma")
ax.set_ylabel("Entropía (bits/n-grama)")
ax.set_title("Comparación de Entropía de fuentes extendidas")
ax.set_xticks(x + width * (len(orders) - 1) / 2)
ax.set_xticklabels(languages)
ax.legend()
ax.grid(True)
plt.show()
